Separate timer and dialog package out from legacy

blmarket opened the pull request:
Hey, I'm not sure it aligns with overall direction you may want, so
just created this PR to explain what I think. Feel free to give your
opinion on it.

Motivation: Encapsulate dialog package so we can have unit test and
integration test available for it without initiating whole
legacy package.

Next steps:

- Remove all memmap usages in dialog package
- Make sure Dialog member fields are never accessed outside of the
  package
- Instantiate Dialog instance

## Required sign-off

- [x] I confirm that my PR does not contain any commercial or protected assets and/or source code.
- [x] I agree in advance that my codes will be licensed automatically under the Apache License or similar BSD/MIT-like
      open source licenses in case if OpenNox Project will adopt such a non-GPL license in the future.

blmarket commented on src/legacy/dialog/dialog.go:
```diff
@@ -93,25 +99,32 @@ func Sub_44D930() bool {
 	return sub_44D930() != 0
 }
 
+func bool2int(b bool) int {
+	if b {
+		return 1
+	}
+	return 0
+}
+
```
XXX: Could not manage to use common library (somehow import broken)

dennwc replied:
Circular dependency, I guess? Or maybe its not exported (name start from lower-case).

blmarket replied:
Hope this small duplication is acceptable given that it looks minor.

dennwc replied:
Sure, no problem.

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -1,11 +1,13 @@
 /*
-#include "defs.h"
-#include "GAME2_2.h"
+#include "../defs.h"
+#include "../GAME2_2.h"
```
This won't work, unfortunately.

Go has a very particular way of handling C build process. C files and headers _must_ be in the same directory. Sub-folders or parent folders won't work - it will just stop recognizing changes to these C files.

This is why we only start moving things to packages when they are completely clean from any C types.

blmarket replied:
Interesting. It worked me (but I experienced longer than usual compile time. Maybe it was compiled twice?)

Anyway, I can simply migrate all C code references using cxgo branch.

dennwc replied:
It kind of works in the sense that it compiles and runs. But it will fail in weird ways afterwards. I will always use cached C file ignoring any future updates to it. So yeah, cxgo port is preferable.

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,248 @@
+	return result
+}
+
+func Sub_486320(a1 unsafe.Pointer, a2 uint32) {
```
Nice! I'd suggest going one small step further.

It looks like `a1` is actually a `*Timer`. So here's what you could do.

Switch to `cxgo` branch (I usually keep a second copy of `opennox` that is always on that branch) and add your `Timer` definition to the `legacy` package. You don't need it to compile.

Then, find `sub_486320` in one of the Go files in that legacy package. Change type of `a1` to your `*Timer`. And do `cd ./src && go run ./internal/fixlegacy`.

It should rewrite this function and replace offsets with actual field accesses. You can then repeat this process for other functions, going from smaller ones to larger ones. I think you can get rid of most `unsafe` accesses here this way.

blmarket replied:
Just saw this suggestion. Let me try it.

blmarket replied:
Yes it did some conversions automatically, but too much additional work jumping two branches. For me it was simply easier to convert them in-place without using cxgo branch.

Can we defer this until we migrate those funcs to member of `Timer`? Wondering we can migrate `Sub_486320` to `func (timer *Timer) SetRaw(target int)` and so on.

blmarket replied:
If the concern is to remove use of unsafe, then how about just apply https://github.com/noxworld-dev/opennox/pull/706/commits/7bef7b80bd6508bb4a54c5c4540be0fccfe79d33 ? most of the changes are covered by unit test.

blmarket commented:
Seems now it goes a bit more wilder than I thought, but just giving a try.

- Replace all C references with `instance()`

dennwc commented on src/legacy/dialog/test_dialog/Dialog/C3MW224K.WAV:
Where's this file coming from? Is it from Nox data? Remember, we cannot distribute any data files in the repository.

blmarket replied:
oops, then I need to replace it with placeholder file.

dennwc commented on src/legacy/dialog/test_dialog/dialog_test.go:
```diff
@@ -0,0 +1,96 @@
+	return uint64(platform.Ticks() / time.Millisecond)
+}
+
+func Setup() {
```
```suggestion
func Setup(t testing.TB) {
```
This way you can call:
```go
t.Cleanup(handles.Release)
```
And it will automatically run this function at the end of the test.

dennwc commented on src/legacy/dialog/test_dialog/dialog_test.go:
```diff
@@ -0,0 +1,96 @@
+	dialog.Sub_43DC40 = func() int { return 0 }
+	dialog.Sub_413890 = func() string { return "" }
+	dialog.GetStringManager = func() *strman.StringManager {
+		mgr := &strman.StringManager{}
+		mgr.ReadJSON("dialog_test.json")
```
```suggestion
		mgr := strman.New()
		if err := mgr.ReadJSON("dialog_test.json"); err != nil {
			t.Fatal(err)
		}
```

dennwc commented on src/legacy/dialog/test_dialog/dialog_test.go:
```diff
@@ -0,0 +1,96 @@
+	dialog.GetStringManager = func() *strman.StringManager {
+		mgr := &strman.StringManager{}
+		mgr.ReadJSON("dialog_test.json")
+		return mgr
```
Looks like you are reading the file each time the function is called instead of doing it once. Maybe worth moving this code into `Setup` and just return the variable here?

dennwc commented on src/legacy/dialog/test_dialog/test_env.go:
```diff
@@ -0,0 +1,76 @@
+package test_dialog
+
+/*
+#include <stdint.h>
+
+uint32_t dword_5d4594_816092;
+uint32_t dword_5d4594_816364;
+uint32_t dword_5d4594_816376;
+uint32_t dword_5d4594_830864;
+uint32_t dword_5d4594_830868;
+char* dword_5d4594_830872;
+char* dword_5d4594_830972;
+uint32_t dword_5d4594_831076;
+uint32_t dword_5d4594_831084;
+uint32_t dword_5d4594_831088;
+uint32_t dword_5d4594_831092;
+uint32_t dword_587000_122856;
+uint32_t dword_587000_122848;
+uint8_t dword_587000_81128_val[96];
+uint8_t dword_587000_122852_val[96];
+void* dword_587000_81128;
+void* dword_587000_122852;
+*/
+import "C"
```
Hmm, I wonder why do you need all of these still. You do not expect it to link against the main `legacy` package, right? Why not make the variables on the Go side then?

blmarket replied:
If we decide to move all the variables to Go side, sure why not.

dennwc replied:
Since you want to wrap this PR while keeping C vars in place, maybe we could at least remove these ones for the test.

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,206 @@
+	return unsafe.Pointer(timerGroup)
+}
+
+func TimerInit(self *Timer, target int32) int32 {
```
```suggestion
func (t *Timer) Init(target int32) int32 {
```
Could make it a method, since its already split from C stuff. Same for other functions that operate on one timer or timer group.

Also note that by Go conventions we do not call the receiver variable `self` or `this` or `timerGroup`. We usually pick a short name like `t` for timer, etc.

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,206 @@
+	return TimerUpdate(self)
+}
+
+func Sub_4862E0(a3 unsafe.Pointer, a4 int32) int32 {
```
Maybe remove this and let other packages call `TimerInit` directly?

blmarket replied:
Then we will lose traceability from the address (if someone tries to search func usage of 4862E0 in the future...)

dennwc replied:
But the `legacy` will still export it to C under the old name, right?

And if not - don't worry, all the old code is still in Git. Could be found if/when needed

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,206 @@
+
+func TimerSetInterp(self *Timer, target uint32) int32 {
+	var a1 unsafe.Pointer = self.C()
+	var (
+		v2 int64
+		v3 int32
+	)
+	*(*uint32)(unsafe.Add(unsafe.Pointer(&v2), 4*0)) = *(*uint32)(unsafe.Add(a1, 8))
+	v3 = int32(*(*uint32)(unsafe.Add(a1, 4)))
+	*(*uint32)(a1) &= 0xFFFFFFFE
+	if uint32(int32(v2)) == uint32(v3) {
+		v2 = int64(PlatformTicks())
+		*(*uint64)(unsafe.Add(a1, 24)) = uint64(v2)
+	}
```
Notice how the code takes address from a local variable `v2`. This is dangerous! Go stack behaves differently to C stack. Go may allocate variables on the heap and then destroy them, if the pointer suddenly disappears (e.g. is cast to `int` type).

I would suggest rewriting this part. I noticed that first 4 bytes of `v2` are used for `if`, and the all 8 bytes are overwritten in the `if` body. Thus, these are two independent variables:
```suggestion
	v2 := *(*uint32)(unsafe.Add(a1, 8))
	v3 := *(*uint32)(unsafe.Add(a1, 4))
	*(*uint32)(a1) &= 0xFFFFFFFE
	if v2 == v3 {
		*(*uint64)(unsafe.Add(a1, 24)) = PlatformTicks()
	}
```

Since it already fixes most of the function body, you could easily fix the rest of unsafe in this function too.

blmarket replied:
Simply missed during migration (haha;;)

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,206 @@
+
+func TimerSetRaw(self *Timer, target uint32) {
+	var a1 unsafe.Pointer = self.C()
+	*(*uint32)(a1) |= 1
+	*(*uint32)(unsafe.Add(a1, 4*2)) = uint32(target) << 16
```
This one is tiny. Worth removing unsafe here for sure!

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,206 @@
+		self.Current = self.Target
+		self.Flags = self.Flags | 2
+		return 1
+	} else {
```
Since `if` branch returns, you could drop `else` and lower the code.

dennwc commented on src/legacy/dialog/dialog.go:
```diff
@@ -0,0 +1,244 @@
+)
+
+type Dialog struct {
+	isEnabled          *uint32     // dword_587000_122848 - Whether dialog component is enabled.
+	volume             *uint32     // dword_5d4594_830868 - stores volume level of dialog at 0-100 range
+	fileToRead         **byte      // dword_5d4594_830872
+	currentPlayingFile **byte      // dword_5d4594_830972
+	currentStream      *ail.Stream // dword_5d4594_831088 - Stores a dialog stream currently playing
+	state              *uint32     // dword_5d4594_830864 - Internal state of dialog component
+	someFlag           *uint32     // dword_5d4594_831084 - Internal state being toggled when loading failed?
+	isFallbackMode     *uint32     // dword_587000_122856 - Initially true, becomes false after first dialog read success
+	isInitialized      *uint32     // dword_5d4594_831076
+	audioDriver        *ail.Driver // dword_5d4594_831092
```
I check the code of the branch, and it looks like you did an awesome job of extracting all functions that use these variables. Congrats! :tada: 

We can completely move these variables to Go side now. Just remove all references from C files (`vardefs.c` most importantly) and switch to local vars instead of pointers here.

blmarket replied:
Not there yet. Still a few functions accessing those fields. For example `sub_44D960`, `sub_44D970`, `sub_44D990`.

Let's keep our scope limited to "make module testable", instead of migrate them all at once.

dennwc commented on src/legacy/timer/timer.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func (t *TimerGroup) Update() {
+	(&t.Timers[0]).Update()
+	(&t.Timers[1]).Update()
+	(&t.Timers[2]).Update()
```
```suggestion
	t.Timers[0].Update()
	t.Timers[1].Update()
	t.Timers[2].Update()
```
As long as `Update` is marked as a pointer method (`func (t *Timer)`) Go will know to use the pointer here.

dennwc reviewed (commented):
OK, lets keep it shorter then. Only two things remain: remove C vars from tests and omit `&` when calling methods in the timer group.

dennwc commented on src/legacy/dialog/test_dialog/test_env.go:
```diff
@@ -0,0 +1,78 @@
+	return GoString(dword_5d4594_830972)
+}
+
+func GoString(p *byte) string { return C.GoString((*C.char)(unsafe.Pointer(p))) }
+func InternCStr(p string) *byte {
+	// Unlike internCStr, we do simpler version.
+	return (*byte)(unsafe.Pointer(C.CString(p)))
+}
```
This still uses C imports in tests. Could you switch to `alloc.InternCString` instead? Same for `GoString`.

dennwc reviewed (commented):
Looks good to me!

Would you allow me to shuffle files around a bit? Since there's no C in tests now, the tests can be in the same package as the dialog, for example.

Also, I got an idea about the `NewDialog`. I was thinking maybe I can move all the function export variables like `Sub_413890` and `GetStringManager` to a function argument of `NewDialog`. Less global state - the better.

blmarket commented:
> Looks good to me!
> 
> Would you allow me to shuffle files around a bit? Since there's no C in tests now, the tests can be in the same package as the dialog, for example.
> 
> Also, I got an idea about the `NewDialog`. I was thinking maybe I can move all the function export variables like `Sub_413890` and `GetStringManager` to a function argument of `NewDialog`. Less global state - the better.

Sure, I can move some by myself, but feel free to take it from here as you want. I'll try to use the same structure for the next work.

dennwc commented:
@blmarket Done! I moved all function to methods on `Dialog` and also removed the global state. Ready to merge, if all is good here.

blmarket commented:
> @blmarket Done! I moved all function to methods on `Dialog` and also removed the global state. Ready to merge, if all is good here.

Looks great to me as well! Thanks for the help! Will try to use similar structure for the future works.
Passed all my manual in-game tests.

dennwc commented:
@blmarket I had to squash all the commits, unfortunately. Initially I wanted to only group them (e.g. squash minor fixes only), but that Nox data file was in the middle of the commit log.
